File: features/features/chat/src/main/java/com/claude/remote/features/chat/ChatViewModel.kt
package com.claude.remote.features.chat

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.claude.remote.core.ssh.DebugLog
import com.claude.remote.core.ssh.SshClient
import com.claude.remote.core.tmux.TmuxSessionManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ChatViewModel — uses JSch SSH connected to Termux TerminalEmulator.
 *
 * SSH connection is handled by JSch (pure Java). Terminal rendering
 * is handled by Termux's TerminalEmulator + TerminalView (native Canvas).
 * SshTerminalSession bridges the two — no external binary needed.
 */
@HiltViewModel
class ChatViewModel @Inject constructor(
    private val sshClient: SshClient,
    val terminalHolder: NativeTerminalHolder,
    private val fileUploadManager: FileUploadManager,
    private val tmuxSessionManager: TmuxSessionManager
) : ViewModel() {

    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()

    init {
        // Keep dot in sync with actual SSH connection state
        viewModelScope.launch {
            sshClient.connectionState.collect { state ->
                _uiState.update { it.copy(connectionState = state) }
            }
        }

        val sessionName = sshClient.currentSessionName
        if (terminalHolder.isSessionRunning() && terminalHolder.attachedSessionName == sessionName) {
            // Same session — reattach existing terminal view
            _uiState.update { it.copy(
                isTerminalMode = true,
                isStreaming = true,
                sessionName = sessionName
            ) }
            terminalHolder.attachExistingSession()
        } else if (sessionName.isNotEmpty()) {
            // New or different session — destroy old and connect fresh
            connectAndAttach(sessionName)
        }
    }

    fun connectAndAttach(sessionName: String) {
        if (sshClient.host.isBlank()) {
            _uiState.update { it.copy(error = "SSH host not configured") }
            return
        }

        terminalHolder.destroySession()

        _uiState.update { it.copy(
            isTerminalMode = true,
            isStreaming = true,
            isReconnecting = true,
            sessionName = sessionName,
            statusMessage = "Connecting..."
        ) }

        viewModelScope.launch {
            try {
                // Ensure SSH is connected before opening terminal channel
                if (sshClient.connectionState.value != com.claude.remote.core.ui.components.ConnectionState.CONNECTED) {
                    DebugLog.log("CHAT", "SshClient not connected, connecting first...")
                    _uiState.update { it.copy(statusMessage = "Connecting to SSH...") }
                    sshClient.connect(sshClient.host, sshClient.port, sshClient.username, sshClient.password)
                }

                // Open new shell channel — retry once on failure (handles stale session)
                _uiState.update { it.copy(statusMessage = "Opening terminal...") }
                try {
                    terminalHolder.createSshSession()
                } catch (e: Exception) {
                    DebugLog.log("CHAT", "First createSshSession failed: ${e.message}, retrying with fresh connection")
                    _uiState.update { it.copy(statusMessage = "Reconnecting SSH...") }
                    sshClient.connect(sshClient.host, sshClient.port, sshClient.username, sshClient.password)
                    terminalHolder.createSshSession()
                }

                // Wait for shell prompt, then attach tmux
                _uiState.update { it.copy(statusMessage = "Attaching to $sessionName...") }
                kotlinx.coroutines.delay(300)
                if (terminalHolder.isSessionRunning() && sessionName.isNotEmpty()) {
                    DebugLog.log("CHAT", "Attaching to tmux session: $sessionName")
                    terminalHolder.writeToSession("tmux attach -t '$sessionName' || tmux new-session -s '$sessionName'\r")
                    terminalHolder.attachedSessionName = sessionName
                    sshClient.currentSessionName = sessionName
                    sshClient.isAttachedToTmux = true
                }
                _uiState.update { it.copy(isReconnecting = false, statusMessage = null) }
            } catch (e: Exception) {
                DebugLog.log("CHAT", "SSH connect failed: ${e.message}")
                _uiState.update { it.copy(
                    error = "SSH failed: ${e.message}",
                    isReconnecting = false,
                    statusMessage = null
                ) }
            }
        }
    }

    fun reconnect() {
        val sessionName = _uiState.value.sessionName
        if (sessionName.isNotEmpty()) {
            connectAndAttach(sessionName)
        }
    }

    /** Probe SSH with a quick command + timeout. If dead or hung, reconnect. Called on resume. */
    fun checkConnectionAndReconnect() {
        val sessionName = _uiState.value.sessionName
        if (sessionName.isEmpty()) return

        _uiState.update { it.copy(statusMessage = "Checking connection...") }

        viewModelScope.launch {
            val wasAttached = sshClient.isAttachedToTmux
            sshClient.isAttachedToTmux = false
            try {
                // Probe with 2s timeout — half-dead SSH may hang instead of throwing
                kotlinx.coroutines.withTimeout(2000) {
                    sshClient.executeCommand("echo ok")
                }
                sshClient.isAttachedToTmux = wasAttached
                _uiState.update { it.copy(statusMessage = null) }
                // SSH alive — just redraw terminal
                terminalHolder.terminalView?.invalidate()
                terminalHolder.writeToSession("\u000c") // Ctrl+L
            } catch (_: Exception) {
                sshClient.isAttachedToTmux = wasAttached
                DebugLog.log("CHAT", "SSH probe failed/timed out on resume, reconnecting")
                connectAndAttach(sessionName)
            }
        }
    }

    fun loadAvailableSessions() {
        _uiState.update { it.copy(isLoadingSessions = true) }
        viewModelScope.launch {
            // Command shell is independent of terminal channel — temporarily
            // allow executeCommand while terminal is attached to tmux
            val wasAttached = sshClient.isAttachedToTmux
            sshClient.isAttachedToTmux = false
            try {
                val sessions = tmuxSessionManager.listSessions(sshClient)
                _uiState.update { it.copy(availableSessions = sessions, isLoadingSessions = false) }
            } catch (e: Exception) {
                DebugLog.log("CHAT", "Failed to load sessions: ${e.message}")
                _uiState.update { it.copy(isLoadingSessions = false) }
            } finally {
                sshClient.isAttachedToTmux = wasAttached
            }
        }
    }

    fun switchSession(sessionName: String) {
        if (sessionName == _uiState.value.sessionName) return
        // Clear old content + update UI immediately so user sees the switch
        terminalHolder.clearScreen()
        terminalHolder.attachedSessionName = sessionName
        sshClient.currentSessionName = sessionName
        _uiState.update { it.copy(sessionName = sessionName) }

        val sshAlive = sshClient.connectionState.value == com.claude.remote.core.ui.components.ConnectionState.CONNECTED
        if (sshAlive && terminalHolder.isSessionRunning()) {
            // Fast path: SSH alive, just switch tmux client
            viewModelScope.launch {
                val wasAttached = sshClient.isAttachedToTmux
                sshClient.isAttachedToTmux = false
                try {
                    sshClient.executeCommand("tmux switch-client -t '${sessionName.replace("'", "'\\''")}'")
                    sshClient.isAttachedToTmux = wasAttached
                } catch (_: Exception) {
                    // Command failed — SSH likely dead, full reconnect
                    sshClient.isAttachedToTmux = wasAttached
                    connectAndAttach(sessionName)
                }
            }
        } else {
            // SSH dead or terminal dead — full reconnect
            connectAndAttach(sessionName)
        }
    }

    fun restartCli(allSessions: Boolean) {
        viewModelScope.launch {
            if (!allSessions) {
                // This session only — send /exit then ccx to current terminal
                terminalHolder.writeToSession("/exit\r")
                kotlinx.coroutines.delay(2000)
                terminalHolder.writeToSession("ccx\r")
                return@launch
            }
            // All sessions
            val wasAttached = sshClient.isAttachedToTmux
            sshClient.isAttachedToTmux = false
            try {
                val sessions = tmuxSessionManager.listSessions(sshClient)
                val sessionNames = sessions.map { it.name }
                DebugLog.log("CHAT", "Restart CLI: ${sessionNames.size} sessions")

                for (name in sessionNames) {
                    val escaped = name.replace("'", "'\\''")
                    sshClient.executeCommand("tmux send-keys -t '$escaped' '/exit' Enter")
                }
                kotlinx.coroutines.delay(2000)
                for (name in sessionNames) {
                    val escaped = name.replace("'", "'\\''")
                    sshClient.executeCommand("tmux send-keys -t '$escaped' 'ccx' Enter")
                }
                sshClient.isAttachedToTmux = wasAttached
            } catch (e: Exception) {
                DebugLog.log("CHAT", "Restart CLI failed: ${e.message}")
                _uiState.update { it.copy(error = "Restart CLI failed: ${e.message}") }
                sshClient.isAttachedToTmux = wasAttached
            }
        }
    }

    fun refreshToken(allSessions: Boolean) {
        viewModelScope.launch {
            if (!allSessions) {
                // This session only — /exit then ca refresh && ccx in current terminal
                terminalHolder.writeToSession("/exit\r")
                kotlinx.coroutines.delay(2000)
                terminalHolder.writeToSession("ca refresh && ccx\r")
                return@launch
            }
            // All sessions
            val wasAttached = sshClient.isAttachedToTmux
            sshClient.isAttachedToTmux = false
            try {
                val sessions = tmuxSessionManager.listSessions(sshClient)
                val sessionNames = sessions.map { it.name }
                DebugLog.log("CHAT", "Refresh token: ${sessionNames.size} sessions to restart")

                for (name in sessionNames) {
                    val escaped = name.replace("'", "'\\''")
                    sshClient.executeCommand("tmux send-keys -t '$escaped' '/exit' Enter")
                }
                kotlinx.coroutines.delay(2000)

                val restartChain = sessionNames.joinToString(" && ") { name ->
                    val escaped = name.replace("'", "'\\''")
                    "tmux send-keys -t '$escaped' 'ccx' Enter"
                }
                val fullCommand = "ca refresh && $restartChain"
                DebugLog.log("CHAT", "Sending refresh chain to terminal: $fullCommand")

                sshClient.isAttachedToTmux = wasAttached
                terminalHolder.writeToSession(fullCommand + "\r")
            } catch (e: Exception) {
                DebugLog.log("CHAT", "Refresh token failed: ${e.message}")
                _uiState.update { it.copy(error = "Refresh token failed: ${e.message}") }
                sshClient.isAttachedToTmux = wasAttached
            }
        }
    }

    fun killSession(sessionName: String) {
        viewModelScope.launch {
            val wasAttached = sshClient.isAttachedToTmux
            sshClient.isAttachedToTmux = false
            try {
                sshClient.executeCommand("tmux kill-session -t '${sessionName.replace("'", "'\\''")}'")
            } catch (_: Exception) {}
            sshClient.isAttachedToTmux = wasAttached
        }
    }

    fun sendRawEscape(sequence: String) {
        terminalHolder.writeBytes(sequence.toByteArray(Charsets.UTF_8))
    }

    /**
     * Send terminal input, but first snap tmux back to the live bottom. When the
     * user has scrolled up, tmux is in copy-mode and would EAT the keystrokes as
     * copy-mode commands (so the typed text vanishes and never reaches Claude).
     * `send-keys -X cancel` exits copy-mode over the command channel (no-op when
     * not in copy-mode, so it's safe either way and injects nothing literal).
     */
    fun sendInput(text: String) {
        val sessionName = _uiState.value.sessionName
        viewModelScope.launch {
            if (sessionName.isNotEmpty()) {
                val escaped = sessionName.replace("'", "'\\''")
                val wasAttached = sshClient.isAttachedToTmux
                sshClient.isAttachedToTmux = false
                try {
                    sshClient.executeCommand("tmux send-keys -t '$escaped' -X cancel 2>/dev/null || true")
                } catch (_: Exception) {}
                sshClient.isAttachedToTmux = wasAttached
            }
            terminalHolder.writeBytes(text.toByteArray(Charsets.UTF_8))
        }
    }

    fun sendMessage() {
        val text = _uiState.value.inputText.trim()
        if (text.isEmpty()) return
        _uiState.update { it.copy(inputText = "") }

        if (_uiState.value.isTerminalMode) {
            terminalHolder.writeToSession(text + "\r")
        }
    }

    fun onInputChanged(text: String) {
        _uiState.update { it.copy(inputText = text) }
    }

    fun uploadAndAttachFile(uri: Uri) {
        // Upload over the command channel (independent of the tmux PTY — no detach
        // needed). The command channel is gated by isAttachedToTmux, so bypass it
        // during upload, then paste the resulting remote path into the terminal.
        viewModelScope.launch {
            _uiState.update { it.copy(isUploading = true, uploadFileName = "uploading...", statusMessage = "Uploading file...") }
            val wasAttached = sshClient.isAttachedToTmux
            sshClient.isAttachedToTmux = false
            try {
                val attachment = fileUploadManager.uploadFile(uri, sshClient)
                sshClient.isAttachedToTmux = wasAttached
                _uiState.update { it.copy(isUploading = false, uploadFileName = null, statusMessage = null) }
                // Type the path into the terminal (trailing space, no newline) so the
                // user can keep typing their prompt around the attached file path.
                attachment.remotePath?.let {
                    terminalHolder.writeBytes(("$it ").toByteArray(Charsets.UTF_8))
                }
            } catch (e: Exception) {
                sshClient.isAttachedToTmux = wasAttached
                _uiState.update {
                    it.copy(isUploading = false, uploadFileName = null, statusMessage = null, error = "Upload failed: ${e.message}")
                }
            }
        }
    }

    fun stopStreaming() {
        _uiState.update { it.copy(isStreaming = false) }
    }

    fun copyMessageContent(messageId: String): String? {
        return _uiState.value.messages.find { it.id == messageId }?.content
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    // Voice input

    private var voiceInputManager: VoiceInputManager? = null

    fun initVoiceInput(context: Context) {
        if (voiceInputManager == null) {
            voiceInputManager = VoiceInputManager(context)
            viewModelScope.launch {
                voiceInputManager?.state?.collect { voiceState ->
                    _uiState.update {
                        it.copy(
                            isVoiceListening = voiceState.isListening,
                            voicePartialResult = voiceState.partialResult
                        )
                    }
                }
            }
        }
    }

    fun toggleVoiceInput() {
        if (_uiState.value.isVoiceListening) {
            voiceInputManager?.stopListening()
        } else {
            voiceInputManager?.startListening { recognizedText ->
                if (_uiState.value.isTerminalMode) {
                    terminalHolder.writeBytes(recognizedText.toByteArray(Charsets.UTF_8))
                } else {
                    _uiState.update {
                        it.copy(inputText = it.inputText + recognizedText)
                    }
                }
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        voiceInputManager?.destroy()
    }
}

File: ios/ClaudeRemote/ClaudeRemote/Features/Chat/ChatViewModel.swift
import SwiftUI
import Combine

/// Ports Android `ChatViewModel.kt` behavior 1:1.
@MainActor
final class ChatViewModel: ObservableObject {
    private let ssh = SshClient.shared
    let holder = TerminalHolder.shared

    @Published var connectionState: ConnectionState = .disconnected
    @Published var sessionName = ""
    @Published var isReconnecting = false
    @Published var statusMessage: String?
    @Published var isUploading = false
    @Published var availableSessions: [TmuxSession] = []
    @Published var isLoadingSessions = false
    @Published var error: String?

    private var cancellables = Set<AnyCancellable>()

    init() {
        ssh.$connectionState
            .receive(on: RunLoop.main)
            .sink { [weak self] in self?.connectionState = $0 }
            .store(in: &cancellables)

        // Swipe on the terminal → tmux copy-mode scroll (no local scrollback in alt buffer).
        holder.onScrollLines = { [weak self] lines in
            self?.scrollHistory(lines: lines)
        }
    }

    /// Scroll exactly like the Android app: with tmux mouse mode on, a finger swipe sends
    /// MOUSE WHEEL events (not arrow keys — those get eaten by the CLI's input as command
    /// history). tmux interprets wheel-up/down as scrollback movement. SGR (1006) format:
    ///   ESC [ < button ; col ; row M   (button 64 = wheel up, 65 = wheel down)
    /// Positive lines = scroll up (into history), negative = down.
    func scrollHistory(lines: Int) {
        guard lines != 0 else { return }
        let up = lines > 0
        let count = min(abs(lines), 8)
        let button = up ? 64 : 65
        // Report the wheel at a fixed cell near the top-middle of the screen.
        let seq = "\u{1b}[<\(button);40;5M"
        let bytes = Array(seq.utf8)
        for _ in 0..<count { holder.writeBytes(bytes) }
    }

    /// Begin: connect SSH (if needed), open terminal channel, attach tmux session.
    func start(sessionName: String) {
        self.sessionName = sessionName
        if holder.isSessionRunning && holder.attachedSessionName == sessionName {
            return // same session already attached
        }
        connectAndAttach(sessionName)
    }

    func connectAndAttach(_ sessionName: String) {
        holder.destroySession()
        self.sessionName = sessionName
        isReconnecting = true
        statusMessage = "Connecting..."

        // Pull connection params from settings (host/port/user may be empty if we
        // navigated here without first connecting on the session screen).
        let settings = SettingsStore.shared
        let host = ssh.host.isEmpty ? settings.sshHost : ssh.host
        let port = ssh.port == 0 ? (Int(settings.sshPort) ?? 22) : ssh.port
        let username = ssh.username.isEmpty ? settings.sshUsername : ssh.username
        let password = ssh.password.isEmpty ? settings.getSshPassword() : ssh.password

        Task {
            do {
                if ssh.connectionState != .connected {
                    statusMessage = "Connecting to SSH..."
                    try await ssh.connect(host: host, port: port,
                                          username: username, password: password)
                }
                statusMessage = "Opening terminal..."
                do {
                    try await holder.createSession(ssh)
                } catch {
                    statusMessage = "Reconnecting SSH..."
                    try await ssh.connect(host: host, port: port,
                                          username: username, password: password)
                    try await holder.createSession(ssh)
                }
                statusMessage = "Attaching to \(sessionName)..."
                try? await Task.sleep(for: .milliseconds(300))
                let esc = sessionName.replacingOccurrences(of: "'", with: "'\\''")
                holder.writeToSession("tmux attach -t '\(esc)' || tmux new-session -s '\(esc)'\r")
                holder.attachedSessionName = sessionName
                ssh.currentSessionName = sessionName
                // Our swipe sends mouse-wheel events; tmux needs mouse mode ON to capture
                // them, and alternate-scroll ON so the wheel scrolls history while a
                // full-screen app (Claude CLI) owns the alt screen. Same as Android.
                try? await Task.sleep(for: .milliseconds(400))
                _ = try? await ssh.executeCommand("tmux set -g mouse on \\; set -g alternate-scroll on")
                ssh.isAttachedToTmux = true
                isReconnecting = false
                statusMessage = nil
            } catch {
                self.error = "SSH failed: \(error.localizedDescription)"
                isReconnecting = false
                statusMessage = nil
            }
        }
    }

    func reconnect() {
        if !sessionName.isEmpty { connectAndAttach(sessionName) }
    }

    /// On resume (app came back to foreground): the terminal PTY channel is usually dead
    /// after the screen was off for a while, even if the command channel still answers —
    /// they're separate channels, so probing with `echo ok` gives a false "alive". Probe
    /// the PTY itself: send Ctrl+L and check we get a redraw; if nothing comes back
    /// quickly, re-attach a fresh PTY (the reliable path).
    func checkConnectionAndReconnect() {
        guard !sessionName.isEmpty else { return }
        Task {
            // Probe the actual PTY: did we receive any bytes shortly after a redraw nudge?
            let gotOutput = await holder.probeRedraw(timeout: 2.0)
            if !gotOutput {
                // PTY is stale → re-attach fresh (same path as session switch).
                connectAndAttach(sessionName)
            }
        }
    }

    func loadAvailableSessions() {
        isLoadingSessions = true
        Task {
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            defer { ssh.isAttachedToTmux = wasAttached }
            do {
                availableSessions = try await TmuxSessionManager.listSessions(ssh)
            } catch { /* keep prior list */ }
            isLoadingSessions = false
        }
    }

    func switchSession(_ name: String) {
        guard name != sessionName else { return }
        // Re-open a fresh PTY attached to the new session. switch-client over a separate
        // command channel doesn't reliably retarget our attached client and left input
        // frozen — re-attaching is a touch slower but always works and keeps input live.
        connectAndAttach(name)
    }

    func restartCli(allSessions: Bool) {
        Task {
            if !allSessions {
                holder.writeToSession("/exit\r")
                try? await Task.sleep(for: .seconds(2))
                holder.writeToSession("ccx\r")
                return
            }
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            defer { ssh.isAttachedToTmux = wasAttached }
            do {
                let names = try await TmuxSessionManager.listSessions(ssh).map(\.name)
                for n in names {
                    let e = n.replacingOccurrences(of: "'", with: "'\\''")
                    _ = try await ssh.executeCommand("tmux send-keys -t '\(e)' '/exit' Enter")
                }
                try? await Task.sleep(for: .seconds(2))
                for n in names {
                    let e = n.replacingOccurrences(of: "'", with: "'\\''")
                    _ = try await ssh.executeCommand("tmux send-keys -t '\(e)' 'ccx' Enter")
                }
            } catch { self.error = "Restart CLI failed: \(error.localizedDescription)" }
        }
    }

    func refreshToken(allSessions: Bool) {
        Task {
            if !allSessions {
                holder.writeToSession("/exit\r")
                try? await Task.sleep(for: .seconds(2))
                holder.writeToSession("ca refresh && ccx\r")
                return
            }
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            do {
                let names = try await TmuxSessionManager.listSessions(ssh).map(\.name)
                for n in names {
                    let e = n.replacingOccurrences(of: "'", with: "'\\''")
                    _ = try await ssh.executeCommand("tmux send-keys -t '\(e)' '/exit' Enter")
                }
                try? await Task.sleep(for: .seconds(2))
                let chain = names.map { n -> String in
                    let e = n.replacingOccurrences(of: "'", with: "'\\''")
                    return "tmux send-keys -t '\(e)' 'ccx' Enter"
                }.joined(separator: " && ")
                ssh.isAttachedToTmux = wasAttached
                holder.writeToSession("ca refresh && \(chain)\r")
            } catch {
                ssh.isAttachedToTmux = wasAttached
                self.error = "Refresh token failed: \(error.localizedDescription)"
            }
        }
    }

    func killSession(_ name: String) {
        Task {
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            try? await TmuxSessionManager.killSession(name, ssh)
            ssh.isAttachedToTmux = wasAttached
        }
    }

    func sendRawEscape(_ seq: String) {
        holder.writeBytes(Array(seq.utf8))
    }

    /// Send terminal input, but first snap tmux back to the live bottom. When the
    /// user has scrolled up, tmux is in copy-mode and would EAT the keystrokes as
    /// copy-mode commands (so the typed text vanishes and never reaches Claude).
    /// `send-keys -X cancel` exits copy-mode over the command channel (no-op when
    /// not in copy-mode, so it's safe either way and injects nothing literal).
    func sendInput(_ text: String) {
        let name = sessionName
        Task { @MainActor in
            if !name.isEmpty {
                let esc = name.replacingOccurrences(of: "'", with: "'\\''")
                let wasAttached = ssh.isAttachedToTmux
                ssh.isAttachedToTmux = false
                _ = try? await ssh.executeCommand("tmux send-keys -t '\(esc)' -X cancel 2>/dev/null || true")
                ssh.isAttachedToTmux = wasAttached
            }
            holder.writeBytes(Array(text.utf8))
        }
    }

    func uploadAndAttachFile(_ url: URL) {
        // Upload over the command channel (independent of the tmux PTY — no detach needed).
        // The command channel is gated by isAttachedToTmux, so bypass it during the upload,
        // then paste the resulting remote path into the terminal for the CLI to read.
        Task {
            isUploading = true
            statusMessage = "Uploading file..."
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            do {
                let remotePath = try await FileUploadManager.upload(url: url, ssh: ssh) { [weak self] p in
                    self?.statusMessage = "Uploading file… \(Int(p * 100))%"
                }
                ssh.isAttachedToTmux = wasAttached
                isUploading = false
                statusMessage = nil
                // Type the path into the terminal (a trailing space, no newline, so the
                // user can keep typing their prompt around the attached file path).
                holder.writeBytes(Array((remotePath + " ").utf8))
            } catch {
                ssh.isAttachedToTmux = wasAttached
                isUploading = false
                statusMessage = nil
                self.error = "Upload failed: \(error.localizedDescription)"
            }
        }
    }
}

/// Runs `op` with a timeout; returns true if it finished without throwing in time.
private func withTimeoutBool(seconds: Double, _ op: @escaping () async throws -> Void) async -> Bool {
    await withTaskGroup(of: Bool.self) { group in
        group.addTask { (try? await op()) != nil }
        group.addTask {
            try? await Task.sleep(for: .seconds(seconds))
            return false
        }
        let result = await group.next() ?? false
        group.cancelAll()
        return result
    }
}
